browser/tor/tor_launcher_factory.cc

```diff
@@ -127,28 +128,172 @@ void TorLauncherFactory::RemoveObserver(tor::TorProfileServiceImpl* service) {
 }
 
 void TorLauncherFactory::OnTorLauncherCrashed() {
-  LOG(ERROR) << "Tor Launcher Crashed";
+  LOG(INFO) << "Tor Launcher Crashed";
   is_starting_ = false;
   for (auto& observer : observers_)
     observer.NotifyTorLauncherCrashed();
 }
 
 void TorLauncherFactory::OnTorCrashed(int64_t pid) {
-  LOG(ERROR) << "Tor Process(" << pid << ") Crashed";
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  LOG(INFO) << "Tor Process(" << pid << ") Crashed";
   is_starting_ = false;
+  is_connected_ = false;
   for (auto& observer : observers_)
     observer.NotifyTorCrashed(pid);
+  KillTorProcess();
+  // Post delayed relaucn for control to stop
+  content::GetUIThreadTaskRunner({})
+    ->PostDelayedTask(FROM_HERE,
+               base::BindOnce(&TorLauncherFactory::RelaunchTor,
+                              weak_ptr_factory_.GetWeakPtr()),
+               base::TimeDelta::FromSeconds(1));
 }
 
 void TorLauncherFactory::OnTorLaunched(bool result, int64_t pid) {
   if (result) {
     is_starting_ = false;
+    // We have to wait for circuit established
+    is_connected_ = false;
     tor_pid_ = pid;
   } else {
     LOG(ERROR) << "Tor Launching Failed(" << pid <<")";
   }
   for (auto& observer : observers_)
     observer.NotifyTorLaunched(result, pid);
+  control_->Start();
+}
+
+void TorLauncherFactory::OnTorControlReady() {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  VLOG(2) << "TOR CONTROL: Ready!";
+  control_->GetVersion(
+      base::BindOnce(&TorLauncherFactory::GotVersion,
+                     weak_ptr_factory_.GetWeakPtr()));
+  control_->GetSOCKSListeners(
+      base::BindOnce(&TorLauncherFactory::GotSOCKSListeners,
+                     weak_ptr_factory_.GetWeakPtr()));
+  control_->Subscribe(tor::TorControlEvent::NETWORK_LIVENESS,
+                      base::DoNothing::Once<bool>());
+  control_->Subscribe(tor::TorControlEvent::STATUS_CLIENT,
+                      base::DoNothing::Once<bool>());
+  control_->Subscribe(tor::TorControlEvent::STATUS_GENERAL,
+                      base::DoNothing::Once<bool>());
+  control_->Subscribe(tor::TorControlEvent::STREAM,
+                      base::DoNothing::Once<bool>());
+}
+
+void TorLauncherFactory::GotVersion(bool error, const std::string& version) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  if (error) {
+    VLOG(1) << "Failed to get version!";
+    return;
+  }
+  VLOG(2) << "Tor version: " << version;
+}
+
+void TorLauncherFactory::GotSOCKSListeners(
+    bool error, const std::vector<std::string>& listeners) {
+  DCHECK_CURRENTLY_ON(BrowserThread::UI);
+  if (error) {
+    VLOG(1) << "Failed to get SOCKS listeners!";
+    return;
+  }
+  VLOG(2) << "Tor SOCKS listeners: ";
```

Comment:
`VLOG_IS_ON`

Reply:
fixed in 12c003763561ef5dc1ef014d4769d24a84c58883